overlord/aspectstate/aspectstate.go

```diff
@@ -0,0 +1,132 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package aspectstate
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/snapcore/snapd/aspects"
+	"github.com/snapcore/snapd/overlord/aspectstate/aspecttest"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func Set(st *state.State, account, bundleName, aspect, field string, value interface{}) error {
+	st.Lock()
+	defer st.Unlock()
+
+	databag, err := getDatabag(st, account, bundleName)
+	if err != nil {
+		if !errors.Is(err, state.ErrNoState) && !errors.Is(err, &aspects.NotFoundError{}) {
+			return err
+		}
+
+		databag = aspects.NewJSONDataBag()
+	}
+
+	accPatterns, err := aspecttest.MockAspect(account, bundleName)
+	if err != nil {
+		return err
+	}
+
+	aspectBundle, err := aspects.NewAspectBundle(bundleName, accPatterns, databag, aspects.NewJSONSchema())
```

Comment:
now that we are starting to use things I'm not completely sure it make sense to attach the databag to the bundle, in theory we could change things to pass it into the Set/Get methods of the aspect? in practice what is really important is how we will deal with this with transactions

Reply:
Yes, we could pass the databag and schema directly into the aspect. I think anyway it seems intuitive for the transaction object to carry those and pass them through when setting and getting data